### PR TITLE
dubbo_proxy: implement dubbo proxy

### PR DESCRIPTION
**Pull request by zyfjeff**
Signed-off-by: tianqian.zyf tianqian.zyf@alibaba-inc.com

*Description*:  Implement dubbo proxy
*Risk Level*: low
*Testing*: unittests
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue] #3998
[Optional *Deprecated*:]

### CODE REVIEW

File: source/common/common/logger.h [hunk 1]
```diff
@@ -47,7 +47,8 @@ namespace Logger {
   FUNCTION(testing)              \
   FUNCTION(thrift)               \
   FUNCTION(tracing)              \
-  FUNCTION(upstream)
+  FUNCTION(upstream)		 \
```
Comment (lizan) [inline source/common/common/logger.h]: nit: alphabetical order

File: source/extensions/filters/network/dubbo_proxy/buffer_helper.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
+#include "extensions/filters/network/dubbo_proxy/buffer_helper.h"
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: Seems this file is pretty similar with `buffer_helper` in thrift_proxy, is there anyway to merge those two? cc @zuercher

Comment (zuercher) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: Someone already started the work of extracting this in #3502, but I'm not sure if that effort is ongoing or not.

File: source/extensions/extensions_build_config.bzl [hunk 3]
```diff
@@ -70,6 +70,7 @@ EXTENSIONS = {
     "envoy.filters.network.redis_proxy":                "//source/extensions/filters/network/redis_proxy:config",
     "envoy.filters.network.tcp_proxy":                  "//source/extensions/filters/network/tcp_proxy:config",
     "envoy.filters.network.thrift_proxy":               "//source/extensions/filters/network/thrift_proxy:config",
+    "envoy.filters.network.dubbo_proxy":               "//source/extensions/filters/network/dubbo_proxy:config",
```
Comment (kimmking) [inline source/extensions/extensions_build_config.bzl]: alphabetical order

File: source/extensions/filters/network/dubbo_proxy/decoder.cc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      }
+
+      decode_ended_ = true;
+      ENVOY_LOG(debug, "dubbo: {} protocol decode ended", protocol_->name());
+    }
+
+    ENVOY_LOG(debug, "dubbo: expected body size is {}", context_.body_size_);
+
+    if (data.length() < context_.body_size_) {
+      ENVOY_LOG(debug, "dubbo: need more data for {} deserialize", serializer_->name());
```
Comment (kimmking) [inline source/extensions/filters/network/dubbo_proxy/decoder.cc]: spell: deserialization

File: test/extensions/filters/network/dubbo_proxy/hessian_serializer_impl_test.cc [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+TEST(HessianProtocolTest, deserializeRpcInvocation) {
+  MockSerializationCallbacks cb;
+  HessianSerializerImpl serializer(cb);
+  // incorrect body size
+  Buffer::OwnedImpl buffer;
+  addSeq(buffer, {
+                     0x05, '2', '.', '0', '.', '2', // Dubbo version
+                     0x04, 't', 'e', 's', 't',      // Service naem
```
Comment (kimmking) [inline test/extensions/filters/network/dubbo_proxy/hessian_serializer_impl_test.cc]: spell: name

**Comment by zyfjeff** [comment]
@lizan I submit a pr to pick up the duplicated code, how about this?

**Comment by yun1129** [comment]
@lizan @zuercher From my point of view I would like to say this PR is really fine-grained. It contains the needed code for making the changes self-contained. Could you please make progress on this?

**Comment by swift1911** [comment]
@zyfjeff Does this PR implement dubbo proxy router like thrift router?

**Comment by swift1911** [comment]
@zyfjeff Does the PR implement Dubbo proxy router like thrift router? In Dubbo proto file it seems that router and cluster is not iimplemented

**Comment by zyfjeff** [comment]
@swift1911  dubbo routing part we are currently designing

File: source/extensions/filters/network/dubbo_proxy/buffer_helper.cc (hunk 2 above)
Comment (zuercher) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: #4344 extracts the commonality.

**Comment by lizan** [comment]
@zyfjeff Can you fix the ASAN test? It is a real fail with undefined behavior:

```
external/envoy/source/extensions/filters/network/dubbo_proxy/hessian_utils.cc:169:29: runtime error: left shift of negative value -8
```

File: source/extensions/filters/network/dubbo_proxy/protocol.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  virtual void onRequestMessage(RequestMessagePtr&& req) PURE;
+  virtual void onResponseMessage(ResponseMessagePtr&& res) PURE;
+};
+
+/**
+ * See https://dubbo.incubator.apache.org/#!/docs/dev/implementation.md?lang=en-us
+ */
+class Protocol {
+public:
+  class Context {
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/protocol.h]: nit: struct?

File: test/extensions/filters/network/dubbo_proxy/utility.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+inline void addSeq(Buffer::Instance& buffer, const std::initializer_list<uint8_t> values) {
+  for (int8_t value : values) {
+    buffer.add(&value, 1);
+  }
+}
+
+inline void addString(Buffer::Instance& buffer, const std::string& s) { buffer.add(s); }
+
+inline std::string bufferToString(Buffer::Instance& buffer) {
```
Comment (lizan) [inline test/extensions/filters/network/dubbo_proxy/utility.h]: why do you need this? Buffer has toString method.

File: source/extensions/filters/network/dubbo_proxy/buffer_helper.cc (hunk 2 above)
Comment (lizan) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: Great, @zyfjeff do you mind waiting a couple days for #4344 and rebase on it?

File: source/extensions/filters/network/dubbo_proxy/filter.cc [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    stats_.cx_destroy_local_with_active_rq_.inc();
+  }
+
+  if (event == Network::ConnectionEvent::LocalClose) {
+    stats_.cx_destroy_remote_with_active_rq_.inc();
+  }
+}
+
+void Filter::onRequestMessage(RequestMessagePtr&& message) {
+  if (!message) {
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/filter.cc]: shouldn't this two be `ASSERT`? i.e. it is certainly a bug in the proxy code so shouldn't be caught but crash.

File: source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    context->is_request_ = true;
+    callbacks_.onRequestMessage(std::move(req));
+  } else {
+    ResponseMessageImplPtr res =
+        std::make_unique<ResponseMessageImpl>(request_id, body_size, is_event);
+    res->fromBuffer(buffer);
+    callbacks_.onResponseMessage(std::move(res));
+  }
+
+  buffer.drain(16);
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc]: drain(MessageSize)?

File: source/extensions/filters/network/dubbo_proxy/protocol.h (hunk 6 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/protocol.h]: done

File: test/extensions/filters/network/dubbo_proxy/utility.h (hunk 7 above)
Comment (zyfjeff) [reply test/extensions/filters/network/dubbo_proxy/utility.h]: done

File: source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc (hunk 9 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.cc]: done

File: source/extensions/filters/network/dubbo_proxy/filter.cc (hunk 8 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/filter.cc]: done

**Comment by zyfjeff** [comment]
@lizan 
> Great, @zyfjeff do you mind waiting a couple days for #4344 and rebase on it?

Yes, i do

File: source/extensions/filters/network/dubbo_proxy/buffer_helper.cc (hunk 2 above)
Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: @lizan  Yes, i do

Comment (lizan) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: #4344 is merged, can you rebase this PR based on them?

Comment (zyfjeff) [reply source/extensions/filters/network/dubbo_proxy/buffer_helper.cc]: @lizan I will take the time to rebase recently.

**Comment by zyfjeff** [comment]
@lizan  You can continue to review.

File: source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+typedef std::unique_ptr<ResponseMessageImpl> ResponseMessageImplPtr;
+
+class DubboProtocolImpl : public Protocol {
+public:
+  DubboProtocolImpl(ProtocolCallbacks& callbacks) : callbacks_(callbacks) {}
+  const std::string& name() const override { return ProtocolNames::get().DUBBO; }
+  virtual bool decode(Buffer::Instance& buffer, Protocol::Context* context) override;
+
+  static const uint8_t MessageSize = 16;
```
Comment (lizan) [inline source/extensions/filters/network/dubbo_proxy/dubbo_protocol_impl.h]: s/const/constexpr/ (throughout)

**Comment by mattklein123** [comment]
@zyfjeff going to go ahead and close this out as it looks like we are making steady progress on the other PRs and this is pretty out of date at this point.

**Pull request closed without merging**